base/shell/explorer/appbar.cpp

```diff
@@ -1,290 +1,466 @@
+ * COPYRIGHT:   Copyright 2008 Vincent Povirk for CodeWeavers
+ *              Copyright 2025 Katayama Hirofumi MZ (katayama.hirofumi.mz@gmail.com)
  */
 
-//
-// Adapted from Wine appbar.c .
-//
-
 #include "precomp.h"
+#include "appbar.h"
 
-#include <wine/list.h>
-
-#define GetPrimaryTaskbar() FindWindowW(L"Shell_TrayWnd", NULL)
-
-struct appbar_cmd
+CAppBarManager::CAppBarManager()
+    : m_hAppBarDPA(NULL)
 {
-    DWORD  dwMsg;
-    ULONG  return_map;
-    DWORD  return_process;
-    struct _AppBarData abd;
-};
+}
 
-struct appbar_response
+CAppBarManager::~CAppBarManager()
 {
-    ULONGLONG result;
-    struct _AppBarData abd;
-};
+    DestroyAppBarDPA();
+}
 
-struct appbar_data
-{
-    struct list entry;
-    HWND hwnd;
-    UINT callback_msg;
-    UINT edge;
-    RECT rc;
-    BOOL space_reserved;
-    /* BOOL autohide; */
-};
-
-static struct list appbars = LIST_INIT(appbars);
-
-static struct appbar_data* get_appbar(HWND hwnd)
+PAPPBAR CAppBarManager::FindAppBar(_In_ HWND hwndAppBar) const
 {
-    struct appbar_data* data;
+    if (!m_hAppBarDPA)
+        return NULL;
 
-    LIST_FOR_EACH_ENTRY(data, &appbars, struct appbar_data, entry)
+    INT nItems = DPA_GetPtrCount(m_hAppBarDPA);
+    while (--nItems >= 0)
     {
-        if (data->hwnd == hwnd)
-            return data;
+        PAPPBAR pAppBar = (PAPPBAR)DPA_GetPtr(m_hAppBarDPA, nItems);
+        if (pAppBar && hwndAppBar == pAppBar->hWnd)
+            return pAppBar;
     }
 
     return NULL;
 }
 
-void appbar_notify_all(HMONITOR hMon, UINT uMsg, HWND hwndExclude, LPARAM lParam)
+void CAppBarManager::EliminateAppBar(_In_ INT iItem)
 {
-    struct appbar_data* data;
-
-    LIST_FOR_EACH_ENTRY(data, &appbars, struct appbar_data, entry)
-    {
-        if (data->hwnd == hwndExclude)
-            continue;
+    LocalFree(DPA_GetPtr(m_hAppBarDPA, iItem));
+    DPA_DeletePtr(m_hAppBarDPA, iItem);
+}
 
-        if (hMon && hMon != MonitorFromWindow(data->hwnd, MONITOR_DEFAULTTONULL))
-            continue;
+void CAppBarManager::DestroyAppBarDPA()
+{
+    if (!m_hAppBarDPA)
+        return;
 
-        SendMessageW(data->hwnd, data->callback_msg, uMsg, lParam);
+    INT nItems = DPA_GetPtrCount(m_hAppBarDPA);
+    while (--nItems >= 0)
+    {
+        ::LocalFree(DPA_GetPtr(m_hAppBarDPA, nItems));
     }
-}
 
-/* send_poschanged: send ABN_POSCHANGED to every appbar except one */
-static void send_poschanged(HWND hwnd)
-{
-    appbar_notify_all(NULL, ABN_POSCHANGED, hwnd, 0);
+    DPA_Destroy(m_hAppBarDPA);
+    m_hAppBarDPA = NULL;
 }
 
-/* appbar_cliprect: cut out parts of the rectangle that interfere with existing appbars */
-static void appbar_cliprect( HWND hwnd, RECT *rect )
+// ABM_NEW
+BOOL CAppBarManager::OnAppBarNew(_In_ const APPBAR_COMMAND *pData)
 {
-    struct appbar_data* data;
-    LIST_FOR_EACH_ENTRY(data, &appbars, struct appbar_data, entry)
+    if (m_hAppBarDPA)
     {
-        if (data->hwnd == hwnd)
+        if (FindAppBar(HWND_FROM_HWND32(pData->hWnd32)))
```

Comment:
You may store this converted HWND into a local variable in the function, instead of invoking `HWND_FROM_HWND32(pData->hWnd32)` everytime.
(+ same suggestions for what's done in the other functions below)